File: scripts/algorithms/M/Most Profit Assigning Work/Most Profit Assigning Work.cpp
// Runtime: 157 ms (Top 36.41%) | Memory: 37.7 MB (Top 62.44%)
class Solution {
public:
    static bool cmp(pair<int,int>a, pair<int,int>b){
        if(a.first == b.first){
            return a.second<b.second;
        }

        return a.first>b.first;
    }
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        int ans = 0;
        vector<pair<int,int>>vp;
        for(int i = 0; i<profit.size(); i++){
            vp.push_back({profit[i], difficulty[i]});
        }
        sort(vp.begin(), vp.end(), cmp);
        sort(worker.begin(), worker.end(), greater<int>());

        //i is for traversing the vp and j is for traversing the worker
        int i = 0, j = 0;

        while(i<vp.size() and j<worker.size()){
            if(worker[j]>=vp[i].second){
                ans+=vp[i].first;
                j++;
            }
            else{
                i++;
            }
        }

        return ans;
    }
};

File: scripts/algorithms/M/Most Profit Assigning Work/Most Profit Assigning Work.js
// Runtime: 171 ms (Top 69.15%) | Memory: 49.2 MB (Top 90.43%)
var maxProfitAssignment = function(difficulty, profit, worker) {
    const data = [];

    for (let i = 0; i < difficulty.length; i++) {
        data.push({ difficulty: difficulty[i], profit: profit[i] });
    }

    data.sort((a, b) => a.difficulty - b.difficulty);

    let maxProfit = 0;

    for (let i = 0; i < data.length; i++) {
        data[i].profit = maxProfit = Math.max(maxProfit, data[i].profit);
    }

    // worker.sort((a, b) => a - b);

    let ans = 0;
    let min = 0;

    for (const skill of worker) {
        let left = 0; // min;
        let right = data.length - 1;

        while (left < right) {
            const mid = Math.floor((left + right + 1) / 2);

            if (data[mid].difficulty > skill) {
                right = mid - 1;
            } else {
                left = mid;
            }
        }

        if (data[left].difficulty > skill) {
            continue;
        }

        // min = left;
        ans += data[left].profit;
    }

    return ans;
};